### PR TITLE
2 - hard

### PR DESCRIPTION
**Pull request by MaksimSvyatov**
Сделал задание на этой ветке. Переработал немного компоненты. Нужно ещё будет добавить collection View.

### CODE REVIEW

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/DetailViewController.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
+//
+//  DetailViewController.swift
+//  EPAM_iOS_School_Practice_2_Star_Wars
+//
+//  Created by Maxim on 18.02.2020.
+//  Copyright © 2020 Maksim Svyatov. All rights reserved.
+//
+
+import UIKit
+import Alamofire
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/DetailViewController.swift]: Alamofire не используется в этом контроллере

Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/DetailViewController.swift]: Поправил

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+//
+
+import Foundation
+import Alamofire
+
+protocol SearchServiceProtocol {
+    func searchPeople(for name: String, completion: @escaping ((People?, Error?)->()))
+}
+
+class SearchService: SearchServiceProtocol {
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift]: Would be better: `NetworkManager` instead `SearchService`

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+//
+//  Created by Maxim on 27.02.2020.
+//  Copyright © 2020 Maksim Svyatov. All rights reserved.
+//
+
+import Foundation
+import Alamofire
+
+protocol SearchServiceProtocol {
+    func searchPeople(for name: String, completion: @escaping ((People?, Error?)->()))
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift]: Use `Result<People, Error>` in completion
https://medium.com/@azamsharp/understanding-result-type-in-swift-5-387d5ef9f45e

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+           .validate()
+           .responseDecodable(of: People.self) { response in
+               if let error = response.error {
+                print("Error: \(error.localizedDescription)")
+                   completion (nil, error)
+                   return
+               }
+            
+            guard let people = response.value else {
+                completion (nil, nil)
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift]: You can create your own `Error` and pass it back. In this case it is not clear what happened. There is nothing, no error, no data.

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        showRecentLabelText()
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        showRecentLabelText()
+        showRecentItems()
+    }
+    
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        guard let count = recentItems?.count else { return 0 }
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift]: Best practice when `return` in new line. 
This approach is more convenient for debugging.

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        guard let destinationVC = segue.destination as? DetailViewController else {
+            return
+        }
+        
+        destinationVC.data = selectedItem  
+    }
+    
+    func showRecentItems() {
+        recentItems = personService.getAllPersons()
+        self.tableView.reloadData()
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift]: `self` no needed

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/DataBase/PersonService.swift [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                    realm.add(person)
+                }
+            } catch {
+                return
+            }
+        }
+    }
+    
+    func save(personList: [Person]) {
+        for person in personList {
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/DataBase/PersonService.swift]: personList.forEach { save(person: $0) }

It's just an alternative.

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Person.swift [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        case name
+        case height
+        case mass
+        case birthYear = "birth_year"
+        case gender
+    }
+}
+
+extension Person: Visualized {
+    var titleLabelText: String {
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Person.swift]: Very very bad approach, strong connection between data.
If there are 4 parameters, what will happen?
Try to use enum, or something of this

```
enum PersonData {
   case height(label: String, value: String)
   case ...
}
```

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Visualized.swift [hunk 9]
```diff
@@ -0,0 +1,14 @@
+//
+//  Visualized.swift
+//  EPAM_iOS_School_Practice_2_Star_Wars
+//
+//  Created by Maxim on 17.02.2020.
+//  Copyright © 2020 Maksim Svyatov. All rights reserved.
+//
+
+protocol Visualized {
+  var titleLabelText: String { get }
```
Comment (a-mkrv) [inline EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Visualized.swift]: Try to rewrite this block with items

**Review by a-mkrv — COMMENTED** [review]
Good job, waiting for corrections 👍

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift (hunk 2 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift]: Fixed

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift (hunk 3 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift]: Fixed

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift (hunk 4 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Networking/SearchService.swift]: Fixed

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift (hunk 5 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift]: Fixed

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift (hunk 6 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/RecentPersonsController.swift]: Fixed

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/DataBase/PersonService.swift (hunk 7 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/DataBase/PersonService.swift]: Fixed with personList.forEach { save(person: $0) }

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Person.swift (hunk 8 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Person.swift]: Fixed, but without enum

File: EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Visualized.swift (hunk 9 above)
Comment (MaksimSvyatov) [reply EPAM_iOS_School_Practice_2_Star_Wars/EPAM_iOS_School_Practice_2_Star_Wars/Model/Visualized.swift]: Fixed

**Review by a-mkrv — APPROVED** [review]

**Pull request merged**